### DISCUSSION TITLE
How to create an "emitter"? (similar to useTaskStore.persist.onFinishHydration)

### DISCUSSION THREAD

**Discussion by mapten** [pmndrs/zustand#1703] [General]
Hi,

I'd like to create an "emitter" (not sure if it is the best definition) function, that behaves like useTaskStore.persist.onFinishHydration:

```
useTaskStore.persist.onFinishHydration(() => {
   // do something
});
```  
  This is the idea:

- Zustand:

```
 type State = {
  foo: string;
  whereMagicHappens; // <-- this
  doSomething: () => void;
}

export const fooStore = create<State>()(
  (set, get) => ({
    foo: '',
    doSomething: () => {
   // do something and then
      whereMagicHappens; // <-- emit
    }
  })
)
```

- React component:

```
const { whereMagicHappens } = fooStore((state) => state);

whereMagicHappens(() => {
  // do something
})
```

Any suggestion? Thanks in advance.

**Comment by JacobWeisenburger** [comment]
What are you trying to "emit"?

Also, I'm trying to recreate your example so I can better understand what you are asking for. But I'm getting an error:

```ts
const fooStore = createStore()(
    ( set, get ) => ( {
        foo: '',
        doSomething: () => {
            whereMagicHappens // Cannot find name 'whereMagicHappens'
        }
    } )
)
```

**Reply by mapten** [reply]
`whereMagicHappens` did mean a pseudocode for a 'something' that I still din't know.

I was checking the source code where `useTaskStore.persist.onFinishHydration` is used. It is using a Publisher / Subscriber pattern.

I've implemented a Publisher / Subscriber in my code. I have a smelly smell that it is not the best approach in Zustand, and I still need to test it (unsubscribe, use a Set to prevent duplicated subscribers, etc).

This is how I've implemented (not real code):

- Zustand:

```
type State = {
  foo: string;
  subscribers: (() => void)[];
  doSomething: () => void;
  publisher: (callback: () => void) => void;
}

export const fooStore = create<State>()(
  (set, get) => ({
    foo: '',
    subscribers: [],
    doSomething: () => {
      // do something and then
      get().listeners.forEach((cb) => cb);
    },
    publisher: (cb: () => void) => {
       set((state) => ({ subscribers: [...state.subscribers, cb] }));
    }
  })
)
```

- Component:

```
const { publisher } = fooStore((state) => state);

publisher(() => {
  // do something
})
```

**Reply by JacobWeisenburger** [reply]
this seems like you are trying to use zustand for something it was not intended to handle.

**Reply by mapten** [reply]
You are probably right. I'll try to move the logic inside a component or custom hook. Thanks.

**Reply by JacobWeisenburger** [reply]
But to be fair, I'm not a zustand expert. So maybe someone else knows the answer.
